File: same_tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def is_same_tree(p, q):
    if not p and not q:
        return True
    if (p and not q) or (q and not p) or p.val != q.val:
        return False
    return is_same_tree(p.left, q.left) and is_same_tree(p.right, q.right)


def test_is_same_tree():
    tree1 = TreeNode(1)
    tree1.left = TreeNode(2)
    tree1.right = TreeNode(3)

    tree2 = TreeNode(1)
    tree2.left = TreeNode(2)
    tree2.right = TreeNode(3)

    print(is_same_tree(tree1, tree2))  # Expected output: True

    tree3 = TreeNode(1)
    tree3.left = TreeNode(2)

    tree4 = TreeNode(1)
    tree4.right = TreeNode(2)

    print(is_same_tree(tree3, tree4))  # Expected output: False

    tree5 = TreeNode(1)
    tree5.left = TreeNode(2)
    tree5.right = TreeNode(1)

    tree6 = TreeNode(1)
    tree6.left = TreeNode(1)
    tree6.right = TreeNode(2)

    print(is_same_tree(tree5, tree6))  # Expected output: False

test_is_same_tree()

File: SameTree.js
const isSameTree = (p, q) => {
    if (!p && !q) return true;
    if ((p && !q) || (q && !p) || p.val !== q.val) return false;
    return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
};


const testIsSameTree = () => {
    
    const tree1 = new TreeNode(1);
    tree1.left = new TreeNode(2);
    tree1.right = new TreeNode(3);

    const tree2 = new TreeNode(1);
    tree2.left = new TreeNode(2);
    tree2.right = new TreeNode(3);

    console.log(isSameTree(tree1, tree2)); // Expected output: true

    const tree3 = new TreeNode(1);
    tree3.left = new TreeNode(2);

    const tree4 = new TreeNode(1);
    tree4.right = new TreeNode(2);

    console.log(isSameTree(tree3, tree4)); // Expected output: false

    const tree5 = new TreeNode(1);
    tree5.left = new TreeNode(2);
    tree5.right = new TreeNode(1);

    const tree6 = new TreeNode(1);
    tree6.left = new TreeNode(1);
    tree6.right = new TreeNode(2);

    console.log(isSameTree(tree5, tree6)); // Expected output: false
}

testIsSameTree();